### ISSUE TITLE
Cross Compile Failed

### ISSUE THREAD

**Issue by SKIYET** [openssl/openssl#9899]
Version Openssl 1.1.1d
System: Debian stretch
Complie tool: mingw-w64
configure: ./Configure mingw64 --cross-compile-prefix=x86_64-w64-mingw32- --prefix=/usr/x86_64-w64-mingw32 shared
result: make passed but make install failed.
log:

```
make depend && make _build_libs
make depend && make _build_engines
make depend && make _build_programs
created directory `C:'
created directory `C:/Program'
created directory `Files'
created directory `Files/Common'
make[1]: Entering directory '/tmp/openssl'
created directory `Files/SSL'
created directory `Files/SSL/certs'
make[1]: Entering directory '/tmp/openssl'
created directory `Files/SSL/private'
created directory `Files/SSL/misc'
make[1]: Entering directory '/tmp/openssl'
install ./apps/CA.pl -> C:/Program Files/Common Files/SSL/misc/CA.pl
cp: target 'Files/SSL/misc/CA.pl.new' is not a directory
Makefile:297: recipe for target 'install_ssldirs' failed
make: *** [install_ssldirs] Error 1
make: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/tmp/openssl'
make[1]: Leaving directory '/tmp/openssl'
make[1]: Leaving directory '/tmp/openssl'
make[1]: Entering directory '/tmp/openssl'
make[1]: Nothing to be done for '_build_libs'.
make[1]: Leaving directory '/tmp/openssl'
make[1]: Entering directory '/tmp/openssl'
make[1]: Nothing to be done for '_build_engines'.
make[1]: Leaving directory '/tmp/openssl'
make[1]: Entering directory '/tmp/openssl'
make[1]: Nothing to be done for '_build_programs'.
make[1]: Leaving directory '/tmp/openssl'
```

**Comment by levitte** [comment]
Yup, there's work going on in #9864 to fix this.

**Comment by levitte** [comment]
As an update, this is kinda sorta the issue that is worked on in #9864, *if you do installation in a mingw environment*...  however, that's not what you're doing, and that got me a little by surprise.
The *expected* action was that installation would be done on the target system.  That being said, I see your prefix configuration option, and wonder why it was so completely ignored.

Will revisit this tomorrow.

**Comment by vszakats** [comment]
Just to note that #9864 wasn't noticed/tested in a Windows mingw environment (it should be a largely similar case though), but `debian:testing` (← docker ← Ubuntu ← AppVeyor) and macOS (Homebrew).

**Comment by bernd-edlinger** [comment]
Theoretically your --prefix should be where the files are located at the target:

` ./Configure mingw64 --cross-compile-prefix=x86_64-w64-mingw32- --prefix="C:/Program Files" shared`

and you install using DESTDIR= where you mount the target file system:
`make install DESTDIR=/mnt/target`

**Comment by SKIYET** [comment]
> 
> 
> As an update, this is kinda sorta the issue that is worked on in #9864, _if you do installation in a mingw environment_... however, that's not what you're doing, and that got me a little by surprise.
> The _expected_ action was that installation would be done on the target system. That being said, I see your prefix configuration option, and wonder why it was so completely ignored.
> 
> Will revisit this tomorrow.

The 1.1.1c edition is ok, but 1.1.1d is illed.
I want to compile aria2c for windows X64 with the mingw64 tool.
Openssl is a dependence for aria2c.

**Comment by SKIYET** [comment]
> 
> 
> Theoretically your --prefix should be where the files are located at the target:
> 
> ` ./Configure mingw64 --cross-compile-prefix=x86_64-w64-mingw32- --prefix="C:/Program Files" shared`
> 
> and you install using DESTDIR= where you mount the target file system:
> `make install DESTDIR=/mnt/target`
I use Debian for cross compiling, the parameter --prefix="C:/Program Files" is wrong.

**Comment by bernd-edlinger** [comment]
How did you configure your 1.1.1c ?

**Comment by SKIYET** [comment]
> 
> 
> How did you configure your 1.1.1c ?

configure: ./Configure mingw64 --cross-compile-prefix=x86_64-w64-mingw32- --prefix=/usr/x86_64-w64-mingw32 shared

It is same.

**Comment by bernd-edlinger** [comment]
> I use Debian for cross compiling, the parameter --prefix="C:/Program Files" is wrong.

NO!!!!  This way your target openssl will have security issues.
It will try to load config and engines from /usr/x86_64-w64-migw32 but have no drive letter.
So it will load rougue files it they are installed there.

**Comment by bernd-edlinger** [comment]
For cross compiling, prefix has to be where the program will be installed at the target.
The path is compiled into the .DLL. if your target uses drive letters, you better specify them.

**Comment by SKIYET** [comment]
> 
> 
> > I use Debian for cross compiling, the parameter --prefix="C:/Program Files" is wrong.
> 
> NO!!!! This way your target openssl will have security issues.
> It will try to load config and engines from /usr/x86_64-w64-migw32 but have no drive letter.
> So it will load rougue files it they are installed there.

Does you regard "C:" or "D:" as a drive letter?
But does debian has a drive letter? Are you sure?
And what does "rougue" meaning?
My cross compilation is on a Debian machine not windows.

**Comment by bernd-edlinger** [comment]
I mean "rogue" i.e. infected with viruses.

**Comment by levitte** [comment]
Ah, I was wondering if an intermediate Unix side staging area was involved in the problem.
This isn't very well supported, it's true...

**Comment by bernd-edlinger** [comment]
There is still an issue, we should be able to configure --prefix="C:/Program Files"
maybe with CONFIGURE_INSIST for cross-compiling.

**Comment by SKIYET** [comment]
> 
> 
> Ah, I was wondering if an intermediate Unix side staging area was involved in the problem.
> This isn't very well supported, it's true...

I am confused why my parameter "--prefix=/usr/x86_64-w64-mingw32" is not useful .
And 1.1.1c is ok.
Because the official document of aria2c told me the openssl can be used to compile aria2c.
I have successfully finished the aria2c compilation with openssl-1.1.1c

**Comment by SKIYET** [comment]
> 
> 
> There is still an issue, we should be able to configure --prefix="C:/Program Files"
> maybe with CONFIGURE_INSIST for cross-compiling.

My debian does not have a "C:/Program Files" directory

**Comment by bernd-edlinger** [comment]
Yes, that is a cross compilation, the path name "C:/Program Files" still need to be known
to the compiled files, which execute on the target, wehre C:/Program Files" exist
and have admin-only access rights.

**Comment by levitte** [comment]
@SKIYET, [`CHANGES`](https://github.com/openssl/openssl/blob/9f8deee2437085a366d16d47352d0e63dae78f4a/CHANGES#L80-L85) gives a quick summery, the rest can be read in the [security advisory](https://www.openssl.org/news/secadv/20190730.txt)

**Comment by SKIYET** [comment]
But I use mingw on debian...do you ask me creat a windows folder and then the point the prefix to this folder？
I cannot understand.

**Comment by bernd-edlinger** [comment]
You do not override the mingw prefix and create a sys-root folder on your linux system
and point DESTDIR= to that folder while make install.
Then you copy the files from the sys-root folder to your mingw system, which should
have C:\Windows, and C:\Program Files and windows O/S

**Comment by SKIYET** [comment]
> 
> 
> You do not override the mingw prefix and create a sys-root folder on your linux system
> and point DESTDIR= to that folder while make install.
> Then you copy the files from the sys-root folder to your mingw system, which should
> have C:\Windows, and C:\Program Files and windows O/S

Does you mean create a folder whose owner is root
And use ./Configure mingw64 --cross-compile-prefix=x86_64-w64-mingw32- --prefix="C:/Program Files" shared
make 
make DESTDIR="sys-root folder".

I use ./Configure mingw64 --cross-compile-prefix=x86_64-w64-mingw32- --prefix="C:/Program Files" shared, but failed. The log is as follow.

Failure!  build file wasn't produced.
Please read INSTALL and associated NOTES files.  You may also have to look over
your available compiler tool chain or change your configuration.

Directory given with --prefix MUST be absolute

**Comment by bernd-edlinger** [comment]
That's why this issue is not closed.
You should remove the --prefix= and let the configure use the default prefix.

**Comment by SKIYET** [comment]
> 
> 
> That's why this issue is not closed.
> You should remove the --prefix= and let the configure use the default prefix.

I think It is difficult to use openssl if I want cross compile a software with linux...
I only want use openssl to compile aria2...

**Comment by SKIYET** [comment]
could you please tell me what should I do if I want create a sys-root folder?

**Comment by petrovr** [comment]
Hi SKIYET,
SKIYET wrote:
>>
>> There is still an issue, we should be able to configure --prefix="C:/Program Files"
>> maybe with CONFIGURE_INSIST for cross-compiling.
> My debian does not have a "C:/Program Files" directory
>
Ignore all kinds of such suggestions.
Use what you want to build/create package, i.e. prefix is not so 
important for binaries.

Read above as later on package installation time to use save location to 
install binaries
Remark: of course do not hard code any drive letter and path as 
suggested! Usually package manager(PM) are able to extract them from 
system settings. Use PM functionality to get them.

After installation on user system binaries may search for configuration 
in "<prefix>" but this is another story and ... (depend from installed 
environment).

Regards,
Roumen Petrov

**Comment by SKIYET** [comment]
> 
> Hi SKIYET, SKIYET wrote:
> > > There is still an issue, we should be able to configure --prefix="C:/Program Files" > maybe with CONFIGURE_INSIST for cross-compiling. My debian does not have a "C:/Program Files" directory
> Ignore all kinds of such suggestions. Use what you want to build/create package, i.e. prefix is not so important for binaries. Read above as later on package installation time to use save location to install binaries Remark: of course do not hard code any drive letter and path as suggested! Usually package manager(PM) are able to extract them from system settings. Use PM functionality to get them. After installation on user system binaries may search for configuration in "<prefix>" but this is another story and ... (depend from installed environment). Regards, Roumen Petrov

Dear

But seemed like there is no openssl-mingw64 in debian if I use APT tool to install openssl.

Thanks

And @levitte ,What shoud I do now?

**Comment by pkreuzt** [comment]
I think my case is a little different, I ran into same issue cross-building to static lib with mingw:

$ LDFLAGS="-static -static-libgcc" ./Configure --cross-compile-prefix=i686-w64-mingw32- no-shared no-zlib no-asm --prefix="$HOME/openssl/install" -static mingw

The problem is the same, it fails on install step. But on a static build, PATH should be irrelevant, since it is only used to search for .a libs when building other dependent binaries.

**Comment by pkreuzt** [comment]
More on this: I have succesfully built some binary against static openssl-1.1.1d, even with failed install step. It seems all libs and includes are in place on specified prefix and the failed message is some cosmetic issue. Not sure this is also the case for shared build, I recommend testing.

**Comment by SKIYET** [comment]
> 
> 
> More on this: I have succesfully built some binary against static openssl-1.1.1d, even with failed install step. It seems all libs and includes are in place on specified prefix and the failed message is some cosmetic issue. Not sure this is also the case for shared build, I recommend testing.

I am sorry but failed.
Due to the configuration progress of aria2 will check whether openssl was installed or not.

**Comment by SKIYET** [comment]
@levitte what can I do now ?
I want to clone the commit to "84eba0d57d22884a81ceda34db108b3d9be8f9c4" mentioned in [ #9864](https://github.com/openssl/openssl/pull/9864) with the command “git checkout 84eba0d57d22884a81ceda34db108b3d9be8f9c4”，it failed.

log is as follow
git checkout 84eba0d57d22884a81ceda34db108b3d9be8f9c4
fatal: reference is not a tree: 84eba0d57d22884a81ceda34db108b3d9be8f9c4

**Comment by SKIYET** [comment]
> 
> 
> More on this: I have succesfully built some binary against static openssl-1.1.1d, even with failed install step. It seems all libs and includes are in place on specified prefix and the failed message is some cosmetic issue. Not sure this is also the case for shared build, I recommend testing.

Dear

could you please tell me the the influence of parameter "no-zlib"?

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev barring further input
